streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java

```diff
@@ -291,14 +314,23 @@ private void putInternal(byte[] binaryKey, byte[] binaryValue) {
     }
 
     private Segment getSegment(long segmentId) {
-        int index = (int) (segmentId % segments.length);
+        if (segmentId > currentSegmentId - segments.length) {
+            int index = (int) (segmentId % segments.length);
 
-        if (segments[index] == null) {
-            segments[index] = new Segment(name + "-" + directorySuffix(segmentId), segmentId);
-            segments[index].openDB(context);
-        }
+            if (segments[index] != null && segments[index].id != segmentId) {
+                cleanup();
```

Comment:
Just a safeguard. With this even when on-disk state store has extraneous segments, ```openExistingSegments``` will clean up such segments. An alternative is move this checking to ```openExistingSegments```, but I'd like to keep it here since the checking is cheap.